GPU Lower Triangular Inverse

SteveBronder opened the pull request:
## Summary

Adds the kernels and functions for performing inversions of lower triangular matrices on the GPU.

At the stan math user level, this PR adds the function `lower_triangular_inverse(A)` that accepts a `matrix_gpu`. Internally, we add 3 kernels which break the inversion into three separate steps following the steps described on page 2 of the Stan OpenCL [paper](https://github.com/SteveBronder/stancon2018/blob/master/report.pdf). `step1` performs a matrix inversion with no blocking on the top lower triangular while `step2` and `step3` Calculates the intermediate and final products needed in the parallel blocked version of the inverse.

## Tests

Tests are available in [`lower_triangular_inverse_test.cpp`](https://github.com/bstatcomp/math/blob/46d21d45a001eb26eccec0773d80280dddc3b898/test/unit/math/gpu/lower_tri_inverse_test.cpp)

1. `inverse_gpu_exception`
- Checks for an `std::invalid_argument` on a non-square matrix_gpu

2. `inverse_gpu_small`
- Checks whether the CPU and GPU version of the matrix inverse have near the same level of precision.

3. `inverse_gpu_big`
- Checks whether the CPU and GPU version of the matrix inverse have near the same level of precision.


## Checklist

- [x] Math issue #1028 

- [x] Copyright holder: (fill in copyright holder information)

By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- Rok Češnovar and Erik Štrumbelj (Faculty of Computer and Information Science, University of Ljubljana)
- Steve Bronder

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

SteveBronder commented:
Unrelated, but inspired by this branch, I'm going to make a pr this weekend to fix the helper macros for accessing the matrices on the gpu. They can be made a lot cleaner if we do something like. 

```cpp
#define A(strideLength, numStrides, offsetLength) A[strideLength * numStrides + offsetLength]
```
We could also have one for 
```cpp
#define A(strideLength, numStrides, offsetLength, numStrides) A[strideLength * numStrides + offsetLength * numStrides]
```

rok-cesnovar commented:
Great idea @SteveBronder 

I think we need a better way of creating input for these tests. 

The dangerous cases we dont want to test with is zeros on the diagonal in the input which would produce NaN values as well as cases where the resulting matrix has extremely large values. While developing this I have seen that the CPU version had values like `1000e+308`, while the maximum double should be 1.7e+308 ( returned by `std::numeric_limits<double>::max()`). The OpenCL version returned `Inf`, which is the expected behaviour.

rok-cesnovar commented:
I did some performance fixes today and am going to do one more pass tomorrow to cleanup all the checks that are not necessary now that we have tracked down the bug with always using default define values.

I am also running the performance tests. I will post some graphs once I gather everything. Spoiler alert, on the V100, this lower triangular inverse is 230x faster than the equivalent mdivide_left_tri on an i7 :) Its around 40x faster on the Titan XP and the GPUs are faster than the i7 for matrices with N>240

rok-cesnovar commented:
These are the speedups that were measured using a separate tests scripts for the gpu -> [link](https://github.com/bstatcomp/math/blob/gpu_performance_tests/test/unit/math/gpu/gpu_performance_test.cpp) and cpu version -> [link](https://github.com/bstatcomp/math/blob/gpu_performance_tests/test/unit/math/gpu/cpu_performance_test.cpp). 

For the CPU I ran the tests on a i7-4790 CPU @ 3.60GHz

Speedup for the Titan XP
![speedup of lower triangular inverse - titan xp](https://user-images.githubusercontent.com/28476796/45591057-7154ce80-b947-11e8-8891-4fe7bf566b97.png)

Speedup for the V100
![speedup of lower triangular inverse - v100](https://user-images.githubusercontent.com/28476796/45591062-8fbaca00-b947-11e8-89e8-49fe6cc53923.png)

Both are improvements of the results I posted on a discourse thread a while ago. Speedup of 1 (GPU time = CPU time) is at N~240 for both GPUs

seantalts commented on test/unit/math/gpu/lower_tri_inverse_test.cpp:
```diff
@@ -0,0 +1,137 @@
+  stan::math::matrix_d m1_cpu(size, size);
+  stan::math::matrix_d m1_cl(size, size);
+
+  m1_cpu = stan::math::mdivide_left_tri<Eigen::Lower>(m1).eval();
```
Is this `eval()` necessary?

rok-cesnovar replied:
No. Its left over because I did performance tests and wanted to force evaluation here for timing purposes. Again cleaning up mistakes...

seantalts commented on test/unit/math/gpu/lower_tri_inverse_test.cpp:
```diff
@@ -0,0 +1,137 @@
+  EXPECT_LT(max_error, 1e-8);
+}
+
+TEST(MathMatrixGPU, inverse_gpu_big_3) {
```
What's with the `inverse_gpu_big_` 1 2 and 3?

rok-cesnovar replied:
Because of the way the parallel algorithm is built I wanted to test a few different sizes. 

A size that is a power of 2 is the cleanest to use in the algoriht, other sizes use padding and such and I wanted to make sure that everything is fine during developing.  I guess we can choose one that is not a power of 2 and leave that one in.

seantalts replied:
Ah no, that's fine. Was just trying to figure out which parts of the code were different and which were the same. Might be nice to create a templated test or define a function that these all use?

rok-cesnovar replied:
Agreed, will make a function that will be called in the 3 tests.

rok-cesnovar commented on stan/math/gpu/kernels/helpers.hpp:
```diff
@@ -15,29 +15,29 @@ std::string helpers =
   #endif
   #ifndef dst
-  #define dst(i,j) dst[j * dst_rows + i]
+  #define dst(i,j) dst[(j) * dst_rows + i]
```
You are right, thanks!

rok-cesnovar commented on stan/math/gpu/kernel_cl.hpp:
```diff
@@ -81,9 +81,9 @@ class kernel_functor {
-    options.insert(base_opts.begin(), base_opts.end());
-    kernel_ = compile_kernel(name, source, options);
-    opts_ = options;
+    base_opts.insert(options.begin(), options.end());
```
The way it was set up before, the default value was never overwritten in cases where the hardware limitations did not allow the default value. 

This meant that on AMD, the 16x16 thread block size was used to spawn the threads, but the kernels were compiled with the THREAD_BLOCK_SIZE 32. 

This switch fixes that. We can look at other solutions if you are not comfortable with this one.

seantalts replied:
it seems roughly equivalent, but I just wanted to double check. When writing the current version I looked up `m.insert()` and saw that it would not overwrite values in `m`. Shouldn't that have the behavior that you're looking for, such that defaults in `base_opts` are not used if they are redefined in `options`?

> Because element keys in a map are unique, the insertion operation checks whether each inserted element has a key equivalent to the one of an element already in the container, and if so, the element is not inserted

(from http://www.cplusplus.com/reference/map/map/insert/)

I would expect this new version to have the behavior that it sounds like you aren't looking for... Unless you are trying to overwrite the `options` with the things in `base_opts` and saying the former are the defaults?

rok-cesnovar replied:
I gave this one a second thought. What we actually need is checking whether the `base_opts` value is smaller than the one in `options`, in that case use the one in `base_opts`, otherwise use the options one. The way I changed it we always use the `base_opts` in which case the `options` have no meaning yeah... 

I will rewrite this.

SteveBronder replied:
Okay I'm pretty sure the above is what we want.

Our default block size is 32, but we also check the whether the sqrt of the `device_thread_block_size` is less than the default. If so than we set the default to the sqrt of the device thread block size. Since that's our max block size we only want to replace it if we have a value that's less. 

@rok-cesnovar do we ever want blocks that are wider than they are long? If so then let's change the default to the full device block size and square blocks can just pass their block size as an option. The only other default options this would effect is `WORK_PER_THREAD` which we can also set as the max block size.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,79 @@
+#ifndef STAN_MATH_GPU_KERNELS_LOWER_TRI_INVERSE_STEP1_HPP
```
I will make a more detailed document that will more precisely explain what is going on in each kernel. You are right regarding the names. I had trouble giving them simple descriptive names. 

The first step is actually a batched lower triangular inverse (we do N inverses of submatrices of size). Using the term batched lower triangular inverse might not be good, because we expect the submatrices to be on the diagonal of the input matrix, so its not a "general" batched inverse.

The second and third step perform special cases of batched matrix multiplication. If you look at slide 6 [here](https://github.com/rok-cesnovar/StanGPU-Workshop/blob/master/slides/Stan_%20GPU.pdf). 
Step 2 and 3 perform the calculation of C3. In the simple case like on slide 6 its 2 matrix multiplications and multiplying with -1. However when we have a larger matrix, like in slide 7. For the matrices marked with (2) we have to perform 4 smaller steps 2 and 3. Thats why I termed it batched.

Hopefully that gives some idea what is done here. Should I just make a Gist or just post it here?

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,79 @@
+     * Calculates the inverse with no blocking
+     *
+     * @param[in,out] A The input matrix.
+     * @param[in, out] V A temporary matrix for storing partial calculations.
```
A GPU kernel can only allocate a matrix in the local memory. We are however very limited by the amount of local memory, which is up to 50x50 double maximum, usually 32x32 doubles.

seantalts replied:
So a kernel can never say "malloc from global memory," is that right? Just checking my understanding of GPU coding.

SteveBronder replied:
> So a kernel can never say "malloc from global memory," is that right? Just checking my understanding of GPU coding.

Yes no malloc exists for kernels

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,79 @@
+      int offset = group * block_size;
+      int i = index;
+      for (int j = 0; j < block_size; j++) {
+        if (i == j) {
```
We spawn `block_size` threads here that populate this `block_size` x `block_size` identity submatrix (each thread is "responsible" for a row). I didnt want to spawn threads  `block_size` x `block_size` threads because only  `block_size` threads are used after this step.

Good catch that this should pause the execution due to the reasons we discussed. This isnt dealt with here for 2 reasons:
- modern GPUs are actually capable of dual-issues (2 instructions per warp). You dont want to abuse this, but we are fine here on most GPUs.
- this is nowhere near the bottleneck of the inverse, so we were more focused on speeding up the step2 and step3

In order to make this more efficient, we could add another kernel that would create the N identity submatrices and would spawn `N`x`block_size`x`block_size` threads. Would you be more comfortable with that? Might make this more readable also.

seantalts replied:
Oh, I really wasn't sure which would happen. I'm still kind of unsure - it seems like if the threads are executing in lock step on the same size blocks (due to padding), then `i` and `j` should basically always be equal at the same time across all threads, right? Thus they might actually take the same branch each time and never pause? 

I was more just trying to understand the execution model better. This code is fine, though now that you mention it, if it's easy enough to separate out subprocedures / subkernels here that might be useful in other kernels, that could be cool.

SteveBronder replied:
> if it's easy enough to separate out subprocedures / subkernels here that might be useful in other kernels, that could be cool

Yes! We can have subkernels and that would make some sections of the code easier to read. I got busy this week, but I want to make another PR to do exactly this. (stackoverflow example [here](https://stackoverflow.com/a/7759844))

rok-cesnovar replied:
I will try to explain this a bit more. Let me know if its helpful. Lets say that we only have to calculate 1 inverse of size 5x5. We would spawn 5 threads in a single thread block (they will also be part of the same warp) with `index` values (which is  `i` in this case) 0,1,2,3,4
The thread with `i` = 0 will set `1.0` on the first loop pass, the thread with `i`=1 on the second pass and so on. 

Subkernels do make this more readable yeah. 

In OpenCL 2.2 a kernel will also be able to spawn new threads (this is called dynamic parallelism). That should be fun :)

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,79 @@
+      barrier(CLK_LOCAL_MEM_FENCE);
+      for (int k = 0; k < block_size; k++) {
+        double factor = A(offset + k, offset + k);
+        int j = index;
```
No, this was not cleaned up... My bad. I was trying a few different versions to see what would work better. Sorry :/

seantalts replied:
No worries! I think something like this might make sense if, as you mentioned for the M, N, K names, i, j, and k are also always used to iterate over those. I think it probably makes sense for us to think of our target audience as people already steeped in linear algebra from the sources that use those naming conventions, if those are the standard. If we're coming up with our own names I would love to be able to tell by looking at it which dimension on which matrix it should be iterating along.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,79 @@
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+      }
+      barrier(CLK_LOCAL_MEM_FENCE);
```
Nope, this one is not needed. Thanks Sean.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,79 @@
+      }
+      barrier(CLK_LOCAL_MEM_FENCE);
+      i = index;
+      for (int j = 0; j < block_size; j++) {
```
No. Again same as before. Poor cleanup job on my part :/ Sorry if this wasted your time...

seantalts replied:
No worries at all, this is what code review is for! I have just been hammered with some silly dev ops stuff, but this is my top priority other than putting out fires like that.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step2.hpp:
```diff
@@ -0,0 +1,122 @@
+     * Calculates the intermediate products in the lower_tri_inverse
+     *
+     * @param[in] A input matrix that is being inverted.
+     * @param[out] C temporary matrix to store the intermediate results.
```
Copy-paste error...

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step2.hpp:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @param[in] A input matrix that is being inverted.
+     * @param[out] C temporary matrix to store the intermediate results.
+     * @param M The number of rows for A.
```
We use the same conventions as in the matrix multiplication https://github.com/stan-dev/math/blob/develop/stan/math/gpu/kernels/matrix_multiply.hpp
because this is actually just a special case of matrix multiply. 

M, N, K are used everywhere in the GPU literature when doing matrix multiplication examples and we then also used it here, I guess a force of habit. But I do agree that a more descriptive name would be better.

seantalts replied:
Hmm, I see. I suppose it's probably better to follow the existing literature. Maybe we can link to a cheat sheet or something in some higher level GPU doc.

seantalts commented on stan/math/gpu/kernels/lower_tri_inverse_step2.hpp:
```diff
@@ -0,0 +1,122 @@
+      const int thread_block_row = get_local_id(0);
+      const int thread_block_col = get_local_id(1);
+      // global thread index
+      const int i = THREAD_BLOCK_SIZE * get_group_id(0) + thread_block_row;
```
I think of `i` and `j` being reserved for iterator variables that keep track of loop indices. Can we give these descriptive names since they're longer-lived and not used that way?

seantalts commented:
Can I help out here? What is being done / what should I take a crack at?

rok-cesnovar commented:
I am preparing a doc (PDF) to help with understanding the kernels. I talked to Steve today and he said he will tackle the remaining stuff. 

I took a few days of for moving, so things got stalled on my end unfortunately.  Sorry about that. Will get back on track in the following days. 

The biggest open tasks are giving the kernels more descriptive names and changing the opts merge as we discussed above. Other than that I think the only open comment is changing `i` and `j` to something more appropriate.

seantalts commented:
No worries! Just wanted to offer help. I can take a look at that though I'm
definitely willing to let the names stay or be whatever is consistent with
the majority of the literature on these matrix operations. So that may be a
no-op, haha.

On Wed, Sep 26, 2018 at 4:02 PM, Rok Češnovar <notifications@github.com>
wrote:

> I am preparing a doc (PDF) to help with understanding the kernels. I
> talked to Steve today and he said he will tackle the remaining stuff.
>
> I took a few days of for moving, so things got stalled on my end
> unfortunately. Sorry about that. Will get back on track in the following
> days.
>
> The biggest open tasks are giving the kernels more descriptive names and
> changing the opts merge as we discussed above. Other than that I think the
> only open comment is changing i and j to something more appropriate.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1030#issuecomment-424850541>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7JU8tj5pey95MVCu-QoOn-Ju7iK1ks5ue91SgaJpZM4Wmezj>
> .
>

SteveBronder commented:
@seantalts the GPU tests have gotten through Jenkins so I think this is okay for a second look.

I think the last bit we have to this is making better names for the step1, step2, and step3.

seantalts commented:
Gotcha! Will take a look tomorrow morning, though I'm also at a conference till Sunday so please bear with me :n

bob-carpenter commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,74 @@
+        } else {
+          V[group * block_size * block_size + j * block_size + index] = 0.0;
+        }
+      }
```
I'd replace this whole loop with

```
std::fill_n(&V[group * block_size * block_size + j * block_size],
            block_size,
            index == j ? 1.0 : 0.0);
```

It will eliminate a lot of the arithmetic and only evaluates `index == j` once.

bob-carpenter commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,74 @@
+      int group = get_group_id(0);
+      int block_size = get_local_size(0);
+      int offset = group * block_size;
+      for (int j = 0; j < block_size; j++) {
```
Storing `group * block_size * block_size + j * block_size` cuts down on code duplication and avoids recomputation.

rok-cesnovar replied:
Changed. Thank you.

bob-carpenter commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,74 @@
+        barrier(CLK_LOCAL_MEM_FENCE);
+        for (int i = k + 1; i < block_size; i++) {
+          factor = A(offset + i, offset + k);
+          if (index < i) {
```
Shouldn't this test be outside the for loop because otherwise the loop has no effect?  `factor` isn't used later.

rok-cesnovar replied:
Fixed.

bob-carpenter commented on stan/math/gpu/kernels/lower_tri_inverse_step2.hpp:
```diff
@@ -0,0 +1,125 @@
+
+      double acc[WORK_PER_THREAD];
+      for (int w = 0; w < WORK_PER_THREAD; w++) {
+        acc[w] = 0.0;
```
This can also just be a `fill_n` call.

seantalts commented:
@bob-carpenter  The GPU kernels are in slightly modified C, not C++ (as far as I'm aware). So no standard library.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,74 @@
+     *  This kernel uses the helper macros available in helpers.cl.
+     */
+    __kernel void lower_tri_inverse_step1(__global double* A,
+                                          __global double* V, int rows) {
```
Change to `tmp_inv`, since these are temporary batches of smaller inverses on the diagonal.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step1.hpp:
```diff
@@ -0,0 +1,74 @@
+      for (int k = 0; k < block_size; k++) {
+        double factor = A(offset + k, offset + k);
+        if (index <= k) {
+          V[group * block_size * block_size + index * block_size + k] /= factor;
```
Renamed it to `tmp_offset`, since this is now used in each access to `tmp`.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step2.hpp:
```diff
@@ -0,0 +1,125 @@
+     *  This kernel uses the helper macros available in helpers.cl.
+     */
+    __kernel void lower_tri_inverse_step2(__global read_only double* A,
+                                          __global double* C, const int M,
```
Agreed. Fixed

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step2.hpp:
```diff
@@ -0,0 +1,125 @@
+
+      // local memory
+      __local double A_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
+      __local double B_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
```
The _local names are changed to C1, C2, C3 labels that are explained in the docs.

rok-cesnovar commented on stan/math/gpu/kernels/lower_tri_inverse_step2.hpp:
```diff
@@ -0,0 +1,125 @@
+        for (int w = 0; w < WORK_PER_THREAD; w++) {
+          const int tiled_i = THREAD_BLOCK_SIZE * tile_ind + thread_block_row;
+          const int tiled_j = THREAD_BLOCK_SIZE * tile_ind + thread_block_col;
+          if ((offset + rows + tiled_j + w * THREAD_BLOCK_SIZE_COL)
```
I tried naming these based on discussion with Steve. Hopefully this helps.

rok-cesnovar commented:
I think I have answered all the review comments. I am attaching a draft version of the docs for these kernels, that will hopefully help understand the kernel code a bit more easily. Feel free to give any comment on the docs also. I will add the step2/step3 memory access explanation tomorrow.

[lower triangular inverse kernels.pdf](https://github.com/stan-dev/math/files/2466669/lower.triangular.inverse.kernels.pdf)

seantalts commented:
Nice, thank you! Will take a look tomorrow!

On Wed, Oct 10, 2018 at 6:13 PM Rok Češnovar <notifications@github.com>
wrote:

> I think I have answered all the review comments. I am attaching a draft
> version of the docs for these kernels, that will hopefully help understand
> the kernel code a bit more easily. Feel free to give any comment on the
> docs also. I will add the step2/step3 memory access explanation tomorrow.
>
> lower triangular inverse kernels.pdf
> <https://github.com/stan-dev/math/files/2466669/lower.triangular.inverse.kernels.pdf>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1030#issuecomment-428751613>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7D8xaNJKTtz4nF4Y90WwFQHM2v_Zks5ujnEGgaJpZM4Wmezj>
> .
>
-- 
📲

rok-cesnovar commented:
Here is the updated doc 
[Lower triangular inverse kernels.pdf](https://github.com/stan-dev/math/files/2470419/Lower.triangular.inverse.kernels.pdf)

I copied everything to a [Google doc](https://docs.google.com/document/d/1hdPdr9dZjeGvI24dbNfnThKz3KdBh1e7WP8WyP0TGgQ/edit?usp=sharing) if anyone wants to comment/edit.

seantalts commented:
Hey, this doc looks great! I'm going to just add a brief section to the top giving context for what this group of kernels is used for and what the broad idea linking them all is. I'm going to borrow heavily from your StanCon talk :)

rok-cesnovar commented:
Great, thanks! I still dont know how to name the steps. Step1 could be something like batch_diagonal_inverse, giving step2 and step3 some descriptive names is a bit harder.

seantalts commented:
I'm going to try to incorporate this doc into the source code so we don't lose track of it...

SteveBronder commented:
Trying to think of some names because I think that + adding the docs will close this out

step 1: `batched_diagonal_lower_triangular_inverse` since it operates on the diagonals

step 2: `left_inverse_multiply` because it does C2[the inverse] * A3

step 3: `negative_left_rectangular_multiply` since
1. C2 * A3 is rectangular
2. the formula is - C2 * A3 * C1

Not great names, but at least semi-informative

rok-cesnovar commented:
I like your idea. My suggestion would be to add the suffix _lower_triangular_inverse since none of these steps are really useful outside the lower triangular inverse. I also added the batched prefix to 2. and 3. since we (mostly) do more than 1 mutliplication. What do you think of these then?

1. `batched_diagonal_lower_triangular_inverse`
2. `batched_left_inverse_multiply_lower_triangular_inverse`
3. `batched_negative_left_rectangular_multiply_lower_triangular_inverse`

rok-cesnovar commented:
They are a bit long though yeah :/

SteveBronder commented:
> diagonal_inverse
> inverse_lower_triangular_multiply
> negative_rectangular_lower_triangular_multiply

I think all the above are technically correct? We can add some details in the docs to explain things the other stuff. I think batch is more of an implementation detail

SteveBronder commented:
I added a few things from the paper and changed the names. I think these names are pretty fine. Rok if you have disagreements on those then I'm very open to changing them, though I think they are an okay mix of shorthand and 'getting the point across'. We should make a wiki page out of the google doc and link to it in each kernel.

SteveBronder commented:
Either Jenkins is down or I goofed up p bad

seantalts commented:
I installed a security update and that has made a lot of the logged out links not work for some reason. Can you try logging in on the home page and then clicking the link?

bob-carpenter commented:
The math library has been inconsistent in how verbose the names
are, but we've been pretty consistent with 

  diagonal    -> diag
  inverse     -> inv
  negative    -> neg
  rectangular -> rect
  triangular  -> tri

which would yield the much less unwieldy

  diag_inv
  inv_lower_tri_multiply
  neg_rect_lower_tri_multiply

SteveBronder commented:
I like those and agree, changing now. Also I'm cleaning up the developer docs and placing it in the stancon2018 GPU repo as a pdf. I think including a link to it in the code docs will be the easiest way to give people who want a deeper understanding of the implementation more background. Is that alright with everyone?

rok-cesnovar commented:
I will take a crack at this. Should I link to the images or just ignore them?

SteveBronder commented:
Oh nice I agree with that. I spent some time tonight cleaning up the google doc a bit, if you have time tmrw to work on this that would be rad!

seantalts commented:
Good question - I think they're really helpful, and it seems easier to link
than to include them in the repo. I'd throw them up on the Stan Math wiki
somewhere and link to them from there.

On Mon, Oct 15, 2018 at 11:58 PM Rok Češnovar <notifications@github.com>
wrote:

> I will take a crack at this. Should I link to the images or just ignore
> them?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1030#issuecomment-430092490>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7FGYQQKJsQ6-PyTZspk7DZ32xuiIks5ulVlMgaJpZM4Wmezj>
> .
>

rok-cesnovar commented:
I have added everything on the wiki, together with the images that will be linked in the code docs. Link: https://github.com/stan-dev/math/wiki/GPU-Kernels

seantalts commented:
Sorry, I meant that I wanted to bring the doc into the code as doc strings. So I'm going to try to move all of the text into the code files that already exist and hopefully generate good Doxygen from that. And I'll have to link to images on the wiki. Sorry I've been so busy - back to back conferences and just gave a talk yesterday, but I'll try to set aside some time tonight.

SteveBronder commented:
Alright so I changed the names to what Bob suggested and added the images and a lot of comments to the kernel code. @seantalts lmk how you feel about this

seantalts commented:
Looks good! I went through and polished the doc a little; I think I probably gave you guys the impression I was looking for more detailed doc than I needed. I was hoping you guys could look at this diff and 1) make sure I didn't mess anything up and 2) get a feel for the level of doc that we're looking for - which stuff I deleted, which stuff I added from the wiki. @SteveBronder already did a pass adding in basically all the relevant stuff from the wiki, so that was great. In general we'd prefer doc to live with the code when possible, we usually don't ask people to doc code that is doing fairly standard stuff (but that is obviously a relative question especially when introducing a new framework to an existing project), and if you can give something a descriptive name and delete a comment that's usually a win. I did this in one place (seemed like 'factor' was less helpful as a name than the comment above it that it was the diagonal element) but I need to double check I did that right.

Here's the diff from my two commits: 
https://github.com/stan-dev/math/pull/1030/files/34bc01405cf8e0997ca52d2029255bdd6e84b871..416b0970a5becda77c6550d0ded2d3a371793b8e

SteveBronder commented on stan/math/gpu/kernels/diag_inv.hpp:
```diff
@@ -0,0 +1,92 @@
+     * of matrix A. More often, TB is smaller than A and A will have lower
+     * triangular inverses calculated on submatrices along the diagonal equal to
+     * the size of the thread block. Running this kernel on a matrix with N = 4
+     * * thread_block will yield a lower triangular matrix with identity
```
```suggestion
     *  thread_block will yield a lower triangular matrix with identity
```

SteveBronder commented:
Looks good besides one little extra star I saw. Thanks!
